### PR TITLE
Valid placement

### PR DESCRIPTION
**Pull request by ajtran303**
I broke down the `valid_placement?` method as much as I could to TDD it, and I did it!
I reworked the code snippet you left for overlapping and wrote a test for it too.
We should be good to move on to working on the render method now.

### CODE REVIEW

File: test/board_test.rb [hunk 1]
```diff
@@ -37,4 +37,133 @@ def test_place_ship
...
+    coords_col = ["A1", "B1", "C1"]
+    coords_diag_1 = ["A1", "B2", "C3"]
+    coords_diag_2 = ["C2", "D3"]
+
+    assert_equal false, board.is_diagonal?(coords_row)
+    assert_equal false, board.is_diagonal?(coords_col)
+    assert_equal true, board.is_diagonal?(coords_diag_1)
+    assert_equal true, board.is_diagonal?(coords_diag_2)
+  end
+
```
Comment (ajtran303) [inline test/board_test.rb:132]: Diagonal == (not a row AND not a column)

File: lib/board.rb [hunk 2]
```diff
@@ -34,11 +34,43 @@ def place(ship, coords)
...
+  end
+
+  def same_length?(ship, coords)
+    ship.length == coords.length
+  end
+
+  def is_diagonal?(coords)
+    !is_row?(coords) && !is_col?(coords)
+  end
+
```
Comment (ajtran303) [inline lib/board.rb:47]: Technically, a diagonal coords array has a "consecutive pattern" 
The program doesn't care about diagonal consecutive patterns
So here, diagonal simply means NOT a row OR a column

File: lib/board.rb [hunk 3]
```diff
@@ -34,11 +34,43 @@ def place(ship, coords)
...
+  def is_row?(coords)
+    coords.all? { |letter| letter[0] == "A"} || coords.all? { |letter| letter[0] == "B"} || coords.all? { |letter| letter[0] == "C"} || coords.all? { |letter| letter[0] == "D"}
+  end
+
+  def is_col?(coords)
+    coords.all? { |num| num[-1] == "1"} || coords.all? { |num| num[-1] == "2"} || coords.all? { |num| num[-1] == "3"} || coords.all? { |num| num[-1] == "4"}
+  end
+
```
Comment (ajtran303) [inline lib/board.rb:55]: These helper methods can maybe be refactored later

File: lib/board.rb [hunk 4]
```diff
@@ -34,11 +34,43 @@ def place(ship, coords)
   end
 
   def valid_placement?(ship, coords)
-  
-    coords.find do |coord|
-      !@cells[coord].empty?
-    end
```
Comment (ajtran303) [inline lib/board.rb:40]: Reworked into helper method on line 72

File: test/board_test.rb [hunk 5]
```diff
@@ -37,4 +37,133 @@ def test_place_ship
...
+  def test_it_knows_if_coords_are_in_a_row
+    board = Board.new
+    coords_true = ["A1", "A2", "A3"]
+    coords_false = ["A1", "B1", "C1"]
+
+    assert_equal true, board.is_row?(coords_true)
+    assert_equal false, board.is_row?(coords_false)
+  end
+
```
Comment (ajtran303) [inline test/board_test.rb:79]: I have tests for every helper method

**Review by ajtran303 — COMMENTED** [review]
I added some comments to help with reviewing.

File: lib/board.rb (hunk 2 above)
Comment (GDemps) [reply lib/board.rb:47]: Nice, good explanation.

File: lib/board.rb (hunk 3 above)
Comment (GDemps) [reply lib/board.rb:55]: Yes, we will take a look once we get everything working.

File: test/board_test.rb (hunk 5 above)
Comment (GDemps) [reply test/board_test.rb:79]: Good testing.

**Comment by GDemps** [comment]
Good job with the valid_placement methods.

**Pull request merged**